### PR TITLE
Add support for Log4j 2's ThreadContext

### LINKED ISSUE #188: Add support for Log4j2's ThreadContext
Similar to SLF4J's MDC, Log4j2 has the concept of a [`ThreadContext`](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/ThreadContext.html) ([Log4j2 manual](https://logging.apache.org/log4j/2.x/manual/thread-context.html)). This ThreadContext acts both as MDC and Nested Diagnostic Context (NDC).

Creating a snapshot should use:
- [`ThreadContext.getContext()`](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/ThreadContext.html#getContext--)
- [`ThreadContext.getImmutableStack()`](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/ThreadContext.html#getImmutableStack--)

Reactivating a snapshot could then use Log4j2's [`CloseableThreadContext`](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/CloseableThreadContext.html) which is able to revert the changes on its own:
1. Call either [`CloseableThreadContext.putAll(...)`](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/CloseableThreadContext.html#putAll-java.util.Map-) or [`CloseableThreadContext.pushAll(...)`](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/CloseableThreadContext.html#pushAll-java.util.List-) (with [`ContextStack.asList()`](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/ThreadContext.ContextStack.html#asList--))
2. Then call either [`CloseableThreadContext.Instance.putAll(...)`](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/CloseableThreadContext.Instance.html#putAll-java.util.Map-) or [`CloseableThreadContext.Instance.pushAll(...)`](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/CloseableThreadContext.Instance.html#pushAll-java.util.List-) (with `ContextStack.asList()`) on the returned `CloseableThreadContext.Instance`

Then when closing the `Context` you can call [`CloseableThreadContext.Instance.close()`](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/CloseableThreadContext.Instance.html#close--) to revert the changes.

The naming of the existing SLF4J MDC library (`mdc-propagation`) might be unfortunate. For Log4j2 I would propose including the library name instead, e.g. `log4j2-propagation`. Also note that just using `log4j` in the name without the "2" might lead to confusion since Log4j1 also has a [MDC](https://logging.apache.org/log4j/1.2/apidocs/org/apache/log4j/MDC.html) and [NDC](https://logging.apache.org/log4j/1.2/apidocs/org/apache/log4j/NDC.html), however Log4j1 is no longer supported so adding support for it here might not be needed.

I just came across your library and have not used it, but if you want I can try creating a pull request.

### PR DESCRIPTION
**Pull request by Marcono1234**
Resolves #188

The code orientates itself to the existing SLF4J MDC one.

Noteworthy:
- Renames the SLF4J MDC module (but not its `artifactId`) to make the distinction between Log4j 1 & 2 (which use the term MDC as well) clearer
- Log4j 2 ThreadContext module currently supports Java 5 even though latest Log4j 2 versions require Java 8; this might allow usage with older Log4j 2 versions, but I have not tested it (relates to #112)
- Log4j 2 `ThreadContextManager.clear()` clears the ThreadContext of the current thread (see also discussion in #204)
- Log4j 2 `ThreadContextManager.getActiveContext().close()` does nothing (similar to SLF4J MDC one)
- Propagating Log4j 2 ThreadContext to a different thread will add the data on top of the existing one instead of replacing it (see also discussion in #202)

### CODE REVIEW

File: log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java
Comment (sjoerdtalsma) [inline log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java:102]: Maybe add _from the same thread_

File: log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java
Comment (sjoerdtalsma) [inline log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java:28]: When multiple context managers are applied it may be useful to know this is Log4j2's data?
Perhaps create an interface `Log4j2ThreadContext` with implementation `Log4j2ThreadcontextData` ?

File: log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java
Comment (sjoerdtalsma) [inline log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java:65]: I prefer `Log4j2ThreadContextManager` in case multiple context managers are used.

File: log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/package-info.java
Comment (sjoerdtalsma) [inline log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/package-info.java:21]: Missing newline at end of file?

File: log4j2-propagation/src/main/resources/META-INF/services/nl.talsmasoftware.context.ContextManager
Comment (sjoerdtalsma) [inline log4j2-propagation/src/main/resources/META-INF/services/nl.talsmasoftware.context.ContextManager:1]: Missing newline at end of file?

File: slf4j-mdc-propagation/pom.xml
Comment (sjoerdtalsma) [inline slf4j-mdc-propagation/pom.xml:30]: Please also rename artifactId to `slf4j-mdc-propagation` to maintain consistency.

We can add a `mdc-propagation` module to the relocations that redirects users to the new artifact.

**Review by sjoerdtalsma — CHANGES_REQUESTED** [review]
Thanks for your great work!
I did an initial review from the online github interface but will check out your branch to have a closer look from the IDE.

Please don't be overwhelmed by the number of comments; I tried to be constructive in my criticism and I like the general quality of this PR, which _again_ is greatly appreciated!

File: log4j2-propagation/README.md
Comment (Marcono1234) [reply log4j2-propagation/README.md:8]: How exactly do you mean that?
Log4j 2 on its own is completely unrelated to SLF4J, it does not implement the SLF4J interface. So someone using only Log4j 2 might not be interested in information about SLF4J.

There is however a [Log4j to SLF4J Adapter](https://logging.apache.org/log4j/2.x/log4j-to-slf4j/index.html) (as well as [Log4j 2 SLF4J Binding](https://logging.apache.org/log4j/2.x/log4j-slf4j-impl/), which is SLF4J to Log4j 2). Do you mean for these 'bridges' a comment would be useful (though I don't actually know how well they work with MDC and ThreadContext)?

File: log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java
Comment (Marcono1234) [reply log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java:38]: These methods are only intended for usage by `ThreadContextManager`. It might also be possible to have this logic only in `ThreadContextManager`, work with the public methods of `ThreadContextData` and have its constructor be package-private, but I felt the logic belongs closer to `ThreadContextData` than to `ThreadContextManager`.

As for why these methods are package-private: Most of them provide internal functionality already provided (indirectly) through `ThreadContextManager` or through Log4j 2's ThreadContext. Providing it publicly might be confusing and lead to incorrect usage.
Making the whole class package-private is not an option because it is publicly exposed through `ThreadContextManager` and would therefore cause usability issues when someone tries to use `ThreadContextManager`.

I don't think making it an interface fits really well because it is not expected that the user implements or subclasses this class. Because the package-private methods are not accessible to the user and are not shown in the javadoc either, I don't think that this is an issue.

Would it be alright for you to leave the class as is? If you want I can add a comment (not part of the documentation) to the class describing the rationale I described here.

Comment (Marcono1234) [reply log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java:38]: There is an instance method `applyToCurrentThread()` as well, which could be `private`, but is package-private for testing (maybe I should rewrite the tests to not use that method though).
The reason why this method exists is similar to what I described in https://github.com/talsma-ict/context-propagation/pull/207#discussion_r577189357, I wanted to have logic related to manipulation of Log4j 2's `ThreadContext` only in `ThreadContextData` because I feel it belongs more there than to `ThreadContextManager`. And because the `data` can be `null`, only having an instance method would not work.

File: log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java
Comment (Marcono1234) [reply log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java:51]: See https://github.com/talsma-ict/context-propagation/pull/207#discussion_r577191621

File: log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java
Comment (Marcono1234) [reply log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java:62]: The methods `getContextMap()` and `getContextStack()` publicly expose the values of these fields and the user should not be able to manipulate them; not even the `ThreadContextData` class itself should be able to manipulate them, since the data represent an unmodifiable snapshot.

It sould also be possible to wrap them in unmodifiable collections every time `getContextMap()` and `getContextStack()` is called, but that would only create unnecessary overhead.

File: log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java
Comment (Marcono1234) [reply log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java:73]: I find it personally easier to work with a `List` than with `ContextStack`.
Additionally opposed to what the name suggests the result of `getImmutableStack()` can actually be manipulated: `ThreadContext.getImmutableStack().asList().add("test")`
I assume that might be a bug, but I have not reported it to Log4j 2 yet. To be on the safe side I rather wanted to rely on the JDK methods for making a list unmodifiable.

Originally I did not expose these fields at all, but then I thought that would make the `ThreadContextData` class look rather useless from outside.
However, in the end I think the user should use Log4j 2's `ThreadContext` directly and not the `ThreadContextData` for accessing the data.

Should I change it anways?

File: log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java
Comment (Marcono1234) [reply log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java:69]: `provider()` is tightly coupled with `ServiceLoader` that I was not sure whether it is good to point the user to that as well. Especially since you normally expect a method like `getInstance()` to return a singleton instance, and not `provider()`.

Do you want this do be changed anyways?

File: log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java
Comment (Marcono1234) [reply log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java:61]: See https://github.com/talsma-ict/context-propagation/pull/207#discussion_r577200546

File: log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java
Comment (Marcono1234) [reply log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java:89]: See https://github.com/talsma-ict/context-propagation/pull/207#discussion_r577200546

File: log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java
Comment (Marcono1234) [reply log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextData.java:28]: I am not sure if having an interface is worth it, see https://github.com/talsma-ict/context-propagation/pull/207#discussion_r577189357. We are not expecting the user to provide their own implementation of that class, are we?

File: slf4j-mdc-propagation/pom.xml
Comment (Marcono1234) [reply slf4j-mdc-propagation/pom.xml:30]: Should we rename it then to `slf4j-propagation` in case this library covers aspects other than MDC in the future as well (this was also the reason why I chose `log4j2-propagation`)?

**Comment by sjoerdtalsma** [comment]
Just to touch base, I have been rather busy lately with many things on my mind.
But let me ensure you, this PR has not been forgotten by me and it's on my list of things to do.

File: slf4j-mdc-propagation/pom.xml
Comment (sjoerdtalsma) [reply slf4j-mdc-propagation/pom.xml:30]: I'd prefer it if the rename were tackled in a separate PR, so there are less different topics to discuss in this one. (I created #209 for the Slf4j module rename)

**Comment by sjoerdtalsma** [comment]
Could you please split this PR into the Log4j2 module only and (optionally) create separate PR's for the other concerns?
It is a bit much for me to review all in a single PR and I would like to be able to focus on the Log4j2 implementation, so we can have a clean first version.

**Comment by sjoerdtalsma** [comment]
I'll create a PR to rename the MDC module (with a new relocation in place for it)

**Comment by Marcono1234** [comment]
> Could you please split this PR into the Log4j2 module only and (optionally) create separate PR's for the other concerns?

Yes, I can do that. Though if you don't mind, it would like to wait for your pull request to be merged first and then rebase this pull request onto `develop`.

File: context-propagation-java5/src/main/java/nl/talsmasoftware/context/ContextSnapshot.java
Comment (Marcono1234) [reply context-propagation-java5/src/main/java/nl/talsmasoftware/context/ContextSnapshot.java:52]: Will be resolved in #217.

File: context-propagation-java5/src/main/java/nl/talsmasoftware/context/clearable/Clearable.java
Comment (Marcono1234) [reply context-propagation-java5/src/main/java/nl/talsmasoftware/context/clearable/Clearable.java:21]: Will be resolved in #217.

File: log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java
Comment (Marcono1234) [reply log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java]: I think it does not hurt to mention this implementation detail; especially since the reason why this method exists is that there is a singleton instance. Otherwise the no-arg constructor would suffice for `ServiceLoader`.

File: log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java
Comment (Marcono1234) [reply log4j2-propagation/src/main/java/nl/talsmasoftware/context/log4j2/ThreadContextManager.java]: Prefixed classes with `Log4j2` and changed package name from `log4j2` to `log4j2.threadcontext` (similar to how you renamed the SLF4J MDC package).

**Comment by Marcono1234** [comment]
Sorry for the delay; I have rebased the changes onto #217 (so that should be merged first), and addressed some of your review comments (which I then marked as resolved).

**Comment by sjoerdtalsma** [comment]
Do you want to rebase on _develop_ again? I merged #217 just now.
I think it would be a good idea to focus on the new `log4j2` module and leave all else out of this PR.

**Comment by sjoerdtalsma** [comment]
I have some issues still, but I can work out those myself after I've merged this.
Let me thank you once more sincerely for your work and your patience with code review 😉.

**Comment by Marcono1234** [comment]
Thanks for merging the changes!

**Comment by sjoerdtalsma** [comment]
> Thanks for merging the changes!

Thanks for contributing them!

**Pull request merged**
